Firestore/core/src/firebase/firestore/model/field_value.cc

```diff
@@ -153,6 +153,72 @@ FieldValue& FieldValue::operator=(FieldValue&& value) {
   }
 }
 
+FieldValue FieldValue::Set(FieldPath field_path, FieldValue value) const {
+  FIREBASE_ASSERT_MESSAGE(type() == Type::Object,
+                          "Cannot set field for non-object FieldValue");
+  FIREBASE_ASSERT_MESSAGE(!field_path.empty(),
+                          "Cannot set field for empty path on FieldValue");
+  const std::string& child_name = field_path.first_segment();
+  if (field_path.size() == 1) {
+    // TODO(zxu): Once immutable type is available, rewrite these.
+    ObjectValue::Map copy = object_value_.internal_value;
+    copy[child_name] = std::move(value);
+    return FieldValue::ObjectValueFromMap(std::move(copy));
+  } else {
+    ObjectValue::Map copy = object_value_.internal_value;
```

Comment:
Why is it necessary to create a call on each recursion?

Reply:
Sorry, a typo destroyed the meaning. I wanted to say "Why is it necessary to create a _copy_ on each recursion?" That is, if I read the code correctly, for nested maps a.b.c, it will create a' = a, then a'.b'  = a.b, then a'.b'.c' = a.b.c. Wouldn't the first copy, a' = a, recursively create copies of b and c? I'm probably missing context here -- i.e., I'm not objecting, just would like some explanation.